controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -475,13 +470,52 @@ func generateKubeadmControlPlaneLabels(clusterName string) map[string]string {
 }
 
 // reconcileDelete handles KubeadmControlPlane deletion.
-func (r *KubeadmControlPlaneReconciler) reconcileDelete(_ context.Context, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (ctrl.Result, error) {
-	err := errors.New("Not Implemented")
+func (r *KubeadmControlPlaneReconciler) reconcileDelete(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (_ ctrl.Result, reterr error) {
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kcp.ObjectMeta)
+	if err != nil {
+		logger.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		logger.Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+	logger = logger.WithValues("cluster", cluster.Name)
 
+	// Fetch Machines
+	allMachines, err := util.GetMachinesForCluster(ctx, r.Client, cluster)
 	if err != nil {
-		logger.Error(err, "Not Implemented")
+		logger.Error(err, "Failed to get list of machines")
 		return ctrl.Result{}, err
 	}
+	ownedMachines := r.getOwnedMachines(kcp, allMachines)
+
+	// Always attempt to update status
+	defer func() {
+		if err := r.updateStatus(ctx, kcp, cluster); err != nil {
+			logger.Error(err, "Failed to update status")
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Verify that only control plane machines remain
+	if !apiequality.Semantic.DeepEqual(allMachines.Items, ownedMachines.Items) {
```

Comment:
Suggestion: compare lengths instead?

Reply:
Good point! I did that initially. Then I changed it, and I can't recall why. I will either fix, or respond with the reason. :grimacing: